server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -469,10 +499,14 @@ private void disassociateDeadNodes(RoutingAllocation allocation) {
             // now, go over all the shards routing on the node, and fail them
             for (ShardRouting shardRouting : node.copyShards()) {
                 final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
-                boolean delayed = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0;
+                boolean delayed = Optional.ofNullable(nodesShutdownMetadata.get(node.nodeId()))
+                    // If we know this node is restarting, then the allocation should be delayed
+                    .map(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+                    // Otherwise, use the "normal" allocation delay logic
+                    .orElse(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexMetadata.getSettings()).nanos() > 0);
```

Comment:
> I think this means that if the shutdown type is REMOVE or REPLACE, we now ignore the delayed left timeout?

Good catch. I'll fix.

> Also, I wonder if we should only use the shutdown indication if it had completed, i.e., the node node was ready to restart when this happened?

We don't actually keep state about whether the shutdown preparation is complete or not, we compute the status when the Get Status API is called. So we'd have to move that logic to somewhere where it's accessible here, and ensure it always remains fast since this is called on the cluster state update thread.

It would likely be better if we checked the status, I agree, but RESTART-type shutdowns should be relatively quick to prepare. Do you think this is something we need for v1, or could we introduce it as the shutdown process becomes more involved?

> A flapping node during a restart could result in a long period of not assigning the shard, since we recalculate the delay every time the shard is recovered and the node subsequently dies.

Can't this happen today with the existing allocation delay? Admittedly, it's less likely with a shorter timeout.

Reply:
There's also the case of someone simply restarting the node before it's ready. Should we use the general delay in that case? It's not clear to me that one behavior is more intuitive than the other.

Reply:
> Do you think this is something we need for v1, or could we introduce it as the shutdown process becomes more involved?

I do not think it is mandatory for v1.

> Can't this happen today with the existing allocation delay? Admittedly, it's less likely with a shorter timeout.

This was specifically for the case where "they setup no delay for their indices". We would be allocating those immediately. Obviously during a cloud restart today, there would be no such allocation going on though.

Given the technical difficulty in doing this, we can defer to a follow-up.

Reply:
> There's also the case of someone intentionally restarting the node before it's ready. Should we use the general delay in that case? It's not clear to me that one behavior is more intuitive than the other.

Thanks, good point. I suppose that for a `RESTART` shutdown, the node is really "ready" for restart immediately (since we always anticipate a restart), however, optimally cloud would wait a little before forcing it through. So relying on the presence of the shutdown indication seems good here.

Reply:
I've changed this to track whether we knew the last allocated node was restarting at the time the shard became unassigned, and if not, to always use the index-level delay rather than the restart-level delay.